### PR TITLE
Upgrade to nvcomp 4.0.1

### PR DESCRIPTION
**Pull request by vuule**
## Description
This PR bumps nvcomp to 4.0.1.

Depends on:
- https://github.com/conda-forge/nvcomp-feedstock/pull/15
- https://github.com/rapidsai/rapids-cmake/pull/633
- https://github.com/rapidsai/kvikio/pull/449

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Review by robertmaynard — CHANGES_REQUESTED** [review]
URL form for nvcomp 4.0 has changed we need `nvcomp-linux-sbsa-4.0.0-cuda12.5.tar.gz` instead of `nvcomp_4.0.0_SBSA_12.x.tgz`

Likewise for x86 which is now `nvcomp-linux-x86_64-4.0.0-cuda12.5.tar.gz`

**Comment by robertmaynard** [comment]
I am going to look more at this given the challenges of figuring out the correct url now for CUDA 11 versus CUDA 12

**Comment by robertmaynard** [comment]
Updated to include the changes from https://github.com/rapidsai/cudf/pull/16250/

**Comment by jakirkham** [comment]
Think we will need PR: https://github.com/conda-forge/nvcomp-feedstock/pull/15

File: cpp/cmake/thirdparty/get_nvcomp.cmake [hunk 1]
```diff
@@ -26,6 +26,7 @@ function(find_and_configure_nvcomp)
   if(TARGET nvcomp AND CUDF_USE_PER_THREAD_DEFAULT_STREAM)
     target_compile_definitions(nvcomp PRIVATE CUDA_API_PER_THREAD_DEFAULT_STREAM)
   endif()
+  target_link_libraries(nvcomp::nvcomp INTERFACE $<LINK_ONLY:CUDA::cuda_driver>)
```
Comment (bdice) [inline cpp/cmake/thirdparty/get_nvcomp.cmake:29]: I think we can remove this with nvcomp 4.0.1, which fixed the hard linkage to `libcuda.so`.

File: java/src/main/native/CMakeLists.txt [hunk 2]
```diff
@@ -267,9 +267,8 @@ if(TARGET nvcomp::nvcomp)
   add_custom_command(
     TARGET cudfjni
     PRE_LINK
-    COMMAND
-      ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> $<TARGET_FILE:nvcomp::nvcomp_gdeflate>
-      $<TARGET_FILE:nvcomp::nvcomp_bitcomp> "${PROJECT_BINARY_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp>
+            "${PROJECT_BINARY_DIR}/libnvcomp.so"
```
Comment (bdice) [inline java/src/main/native/CMakeLists.txt:271]: Do we need to specify the library name? What changed relative to the old code?

File: rapids_config.cmake [hunk 3]
```diff
@@ -12,6 +12,7 @@
 # the License.
 # =============================================================================
 file(READ "${CMAKE_CURRENT_LIST_DIR}/VERSION" _rapids_version)
+
```
Comment (bdice) [inline rapids_config.cmake:15]: Revert this extra newline.
```suggestion
```

**Comment by bdice** [comment]
https://github.com/conda-forge/nvcomp-feedstock/pull/15 is now merged. We'll be able to test this adequately in CI now.

Is this update targeting 24.10? I thought I might've heard it was targeting 24.12 but I'm not sure.

File: cpp/cmake/thirdparty/get_nvcomp.cmake (hunk 1 above)
Comment (robertmaynard) [reply cpp/cmake/thirdparty/get_nvcomp.cmake:29]: Yes we can

File: java/src/main/native/CMakeLists.txt (hunk 2 above)
Comment (robertmaynard) [reply java/src/main/native/CMakeLists.txt:271]: good catch. We don't need to specify the library name

File: cpp/cmake/thirdparty/get_nvcomp.cmake [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # =============================================================================
-# Copyright (c) 2021-2022, NVIDIA CORPORATION.
+# Copyright (c) 2021-2024, NVIDIA CORPORATION.
```
Comment (bdice) [inline cpp/cmake/thirdparty/get_nvcomp.cmake:2]: No changes in this file -- no copyright change needed.
```suggestion
# Copyright (c) 2021-2022, NVIDIA CORPORATION.
```

**Comment by bdice** [comment]
I fixed style checks so that CI can run. I'd like to see if conda builds work now.

File: rapids_config.cmake [hunk 5]
```diff
@@ -33,4 +33,8 @@ if(NOT EXISTS "${CMAKE_CURRENT_BINARY_DIR}/CUDF_RAPIDS-${RAPIDS_VERSION_MAJOR_MI
     "${CMAKE_CURRENT_BINARY_DIR}/CUDF_RAPIDS-${RAPIDS_VERSION_MAJOR_MINOR}.cmake"
   )
 endif()
+
+set(rapids-cmake-repo vuule/rapids-cmake)
+set(rapids-cmake-branch test-nvcomp-4.0)
```
Comment (vuule) [inline rapids_config.cmake:38]: REMOVE BEFORE MERGING

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
I still found `libnvcomp_gdeflate.so` and `libnvcomp_bitcomp.so` in many other places. Should we remove them all?

File: java/src/main/native/CMakeLists.txt [hunk 6]
```diff
@@ -267,9 +267,7 @@ if(TARGET nvcomp::nvcomp)
   add_custom_command(
     TARGET cudfjni
     PRE_LINK
-    COMMAND
-      ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> $<TARGET_FILE:nvcomp::nvcomp_gdeflate>
-      $<TARGET_FILE:nvcomp::nvcomp_bitcomp> "${PROJECT_BINARY_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> "${PROJECT_BINARY_DIR}"
```
Comment (ttnghia) [inline java/src/main/native/CMakeLists.txt:270]:
```suggestion
    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> "${PROJECT_BINARY_DIR}/libnvcomp.so""
```

File: java/src/main/native/CMakeLists.txt [hunk 7]
```diff
@@ -267,9 +267,7 @@ if(TARGET nvcomp::nvcomp)
   add_custom_command(
     TARGET cudfjni
     PRE_LINK
-    COMMAND
-      ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> $<TARGET_FILE:nvcomp::nvcomp_gdeflate>
-      $<TARGET_FILE:nvcomp::nvcomp_bitcomp> "${PROJECT_BINARY_DIR}"
+    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> "${PROJECT_BINARY_DIR}/libnvcomp.so""
```
Comment (ttnghia) [inline java/src/main/native/CMakeLists.txt:270]:
```suggestion
    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:nvcomp::nvcomp> "${PROJECT_BINARY_DIR}/libnvcomp.so"
```

**Review by bdice — APPROVED** [review]
Looks good to me. We will need a final approval from @robertmaynard since he requested changes earlier. (Also, this is a reminder to revert the rapids-cmake branch changes before merging.)

**Pull request merged**